pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -404,7 +427,10 @@ func adsPushAll() {
 	// TODO: get service, serviceinstances, configs once, to avoid repeated redundant calls.
 	// TODO: indicate the specific events, to only push what changed.
 	for _, client := range tmpMap {
-		client.pushChannel <- &XdsEvent{}
+		select {
+		case client.pushChannel <- &XdsEvent{}:
+		case <-client.doneChannel:
```

Comment:
add default case to detect non-blocking write failure.
Keep track of last successful push to the client. Log an error message if last push was more than `timeout` ago, and close the client connection.